File: homework-1/backend-go/internal/store/store.go
package store

import (
	"sync"

	"bankingapi/internal/model"
)

// TransactionStore is an in-memory, thread-safe store (no database).
type TransactionStore struct {
	mu    sync.RWMutex
	items map[string]model.Transaction
}

func New() *TransactionStore {
	return &TransactionStore{items: make(map[string]model.Transaction)}
}

func (s *TransactionStore) Save(t model.Transaction) model.Transaction {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.items[t.ID] = t
	return t
}

func (s *TransactionStore) FindAll() []model.Transaction {
	s.mu.RLock()
	defer s.mu.RUnlock()
	out := make([]model.Transaction, 0, len(s.items))
	for _, t := range s.items {
		out = append(out, t)
	}
	return out
}

func (s *TransactionStore) FindByID(id string) (model.Transaction, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	t, ok := s.items[id]
	return t, ok
}

// Clear removes all transactions (used to isolate tests).
func (s *TransactionStore) Clear() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.items = make(map[string]model.Transaction)
}

File: homework-1/backend-fastapi/app/store.py
import threading
from typing import Optional

from .models import Transaction


class TransactionStore:
    """In-memory, thread-safe transaction store (no database)."""

    def __init__(self) -> None:
        self._items: dict[str, Transaction] = {}
        self._lock = threading.Lock()

    def save(self, transaction: Transaction) -> Transaction:
        with self._lock:
            self._items[transaction.id] = transaction
        return transaction

    def find_all(self) -> list[Transaction]:
        with self._lock:
            return list(self._items.values())

    def find_by_id(self, transaction_id: str) -> Optional[Transaction]:
        with self._lock:
            return self._items.get(transaction_id)

    def clear(self) -> None:
        with self._lock:
            self._items.clear()
